cirq-google/cirq_google/study/zip_longest.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Dict, Iterator, List
+
+import itertools
+import cirq
+
+
+class ZipLongest(cirq.Zip):
+    """Iterate over constituent sweeps in parallel
+
+    Analogous to itertools.zip_longest.
+    Note that we iterate until all sweeps terminate,
+    so if the sweeps are different lengths, the
+    shorter sweeps will be filled by repeating their last value
+    until all sweeps have equal length.
+    This is different from itertools.zip_longest, which uses a fixed fill value.
+    """
+
+    def __init__(self, *sweeps: cirq.Sweep) -> None:
+        self.sweeps = sweeps
+
+    def __eq__(self, other):
+        if not isinstance(other, ZipLongest):
+            return NotImplemented
+        return self.sweeps == other.sweeps
+
+    def __hash__(self) -> int:
+        return hash(tuple(self.sweeps))
+
+    @property
+    def keys(self) -> List['cirq.TParamKey']:
+        return sum((sweep.keys for sweep in self.sweeps), [])
+
+    def __len__(self) -> int:
+        if not self.sweeps:
+            return 0
+        return max(len(sweep) for sweep in self.sweeps)
+
+    def __repr__(self) -> str:
+        sweeps_repr = ', '.join(repr(s) for s in self.sweeps)
+        return f'cirq_google.ZipLongest({sweeps_repr})'
+
+    def __str__(self) -> str:
+        sweeps_repr = ', '.join(repr(s) for s in self.sweeps)
+        return f'ZipLongest({sweeps_repr})'
+
+    def param_tuples(self) -> Iterator[cirq.study.sweeps.Params]:
+        iters = [
+            itertools.chain(sweep.param_tuples(), itertools.repeat(list(sweep.param_tuples())[-1]))
+            for sweep in self.sweeps
+        ]
+        for vals in itertools.islice(zip(*iters), len(self)):
+            yield sum(vals, ())
```

Comment:
This has bitten me before - `sum(iterable, ())` can be very small due to allocation and deallocation of intermediate tuples - consider using `tuple(itertools.chain.from_iterable(vals))` instead
```
In [1]: import itertools
In [2]: tlist = [(i,) for i in range(5000)]
In [3]: %timeit t0 = sum(tlist, ())
17.9 ms ± 182 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [4]: %timeit t1 = tuple(itertools.chain.from_iterable(tlist))
153 µs ± 1.7 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

Reply:
Or an explicit comprehension:
```python
(item for val in vals for item in val)
```

Reply:
Done.